File: app/src/main/java/com/sstechcanada/todo/activities/AddCategoryActivity.java
package com.sstechcanada.todo.activities;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.AppCompatImageView;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.sstechcanada.todo.R;
import com.sstechcanada.todo.activities.auth.LoginActivity;
import com.sstechcanada.todo.adapters.CategoryAdapter;
import com.sstechcanada.todo.models.Category;

import java.util.ArrayList;
import java.util.List;

import es.dmoral.toasty.Toasty;

public class AddCategoryActivity extends AppCompatActivity {

    //view objects
    EditText editTextName;
    Button buttonAddCategory;
    ListView listViewCategory;
    TextView toolBarTitle;
    private AppCompatImageView toolbar_profile;

    List<Category> categories;

    DatabaseReference databaseCategories;
    String userID;
    private FirebaseAuth mAuth;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_category);

        mAuth = FirebaseAuth.getInstance();
        userID = mAuth.getCurrentUser().getUid();

//        databaseCategories = FirebaseDatabase.getInstance().getReference("categories");
        databaseCategories = FirebaseDatabase.getInstance().getReference(userID).child("benefits");

        editTextName = (EditText) findViewById(R.id.editTextName);
        listViewCategory = (ListView) findViewById(R.id.listViewCategory);

        buttonAddCategory = (Button) findViewById(R.id.buttonAddCategory);

        categories = new ArrayList<>();

        toolBarTitle = findViewById(R.id.toolbarTitle);
        toolBarTitle.setText("Add Benefits");

        toolbar_profile = findViewById(R.id.profile_toolbar);
        toolbar_profile.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startActivity(new Intent(AddCategoryActivity.this, LoginActivity.class));
            }
        });

        buttonAddCategory.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                addCategory();
            }
        });

        listViewCategory.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {

                Category category = categories.get(i);
                showUpdateDialog(category.getCategoryId(), category.getCategoryName());

//                //creating an intent
//                Intent intent = new Intent(getApplicationContext(), CategoryActivity.class);
//
//                //putting category name and id to intent
//                intent.putExtra(CATEGORY_ID, category.getCategoryId());
//                intent.putExtra(CATEGORY_NAME, category.getCategoryName());
//
//                //starting the activity with intent
//                startActivity(intent);
            }
        });

        listViewCategory.setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {
            @Override
            public boolean onItemLongClick(AdapterView<?> adapterView, View view, int i, long l) {
                Category category = categories.get(i);
                showUpdateDialog(category.getCategoryId(), category.getCategoryName());
                return true;
            }
        });
    }

    /*
     * This method is saving a new category to the
     * Firebase Realtime Database
     * */
    private void addCategory() {
        //getting the values to save
        String name = editTextName.getText().toString().trim();

        //checking if the value is provided
        if (!name.isEmpty()) {

            //getting a unique id using push().getKey() method
            //it will create a unique id and we will use it as the Primary Key for our Category
            String id = databaseCategories.push().getKey();

            //Making first word capital
            name = name.substring(0, 1).toUpperCase() + name.substring(1);
            //creating an Category Object
            Category category = new Category(id, name);

            //Saving the Category
            databaseCategories.child(id).setValue(category);

            //setting edittext to blank again
            editTextName.setText("");

            //displaying a success toast
            Toasty.success(this, "Benefit added", Toast.LENGTH_SHORT).show();
        } else {
            //if the value is not given displaying a toast
            Toasty.warning(this, "Please enter a name", Toast.LENGTH_SHORT).show();
        }
    }


    @Override
    protected void onStart() {
        super.onStart();
        //attaching value event listener
        databaseCategories.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {

                //clearing the previous category list
                categories.clear();

                //iterating through all the nodes
                for (DataSnapshot postSnapshot : dataSnapshot.getChildren()) {
                    Category category = postSnapshot.getValue(Category.class);
                    categories.add(category);
                }

                //creating adapter
                CategoryAdapter categotyAdapter = new CategoryAdapter(AddCategoryActivity.this, categories);
                //attaching adapter to the listview
                listViewCategory.setAdapter(categotyAdapter);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }

    private void showUpdateDialog(final String categoryId, String categoryName) {

        AlertDialog.Builder dialogBuilder = new AlertDialog.Builder(this);
        LayoutInflater inflater = getLayoutInflater();
        final View dialogView = inflater.inflate(R.layout.update_dialog, null);
        dialogBuilder.setView(dialogView);

        final EditText editTextName = (EditText) dialogView.findViewById(R.id.editTextName);
        final Button buttonUpdate = (Button) dialogView.findViewById(R.id.buttonUpdateCategory);
        final Button buttonDelete = dialogView.findViewById(R.id.buttonDeleteCategory);


        dialogBuilder.setTitle(categoryName);
        final AlertDialog b = dialogBuilder.create();
        b.show();


        buttonUpdate.setOnClickListener(view -> {
            String name = editTextName.getText().toString().trim();
            if (!TextUtils.isEmpty(name)) {
                updateCategory(categoryId, name);
                b.dismiss();
            }
        });

        buttonDelete.setOnClickListener(view -> {

            deleteCategory(categoryId);
            b.dismiss();
        });

    }


    private void updateCategory(String id, String name) {
        //getting the specified category reference
        DatabaseReference dR = FirebaseDatabase.getInstance().getReference(userID).child("benefits").child(id);
//        DatabaseReference dR = FirebaseDatabase.getInstance().getReference("categories").child(id);
        //updating category
        Category category = new Category(id, name);
        dR.setValue(category);
        Toasty.success(getApplicationContext(), "Benefits Updated", Toast.LENGTH_SHORT).show();
    }


    private void deleteCategory(String id) {
        //getting the specified category reference
        DatabaseReference dR = FirebaseDatabase.getInstance().getReference(userID).child("benefits").child(id);
//        DatabaseReference dR = FirebaseDatabase.getInstance().getReference("categories").child(id);
        //removing category
        dR.removeValue();

        //getting the tracks reference for the specified category
        Toasty.error(getApplicationContext(), "Benefits Deleted", Toast.LENGTH_SHORT).show();

    }


}

File: app/src/main/java/com/sstechcanada/todo/activities/AddCategoryActivity.kt
package com.sstechcanada.todo.activities

import androidx.appcompat.app.AppCompatActivity
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.CollectionReference
import com.sstechcanada.todo.adapters.CategoryFirestoreAdapter
import android.os.Bundle
import com.sstechcanada.todo.R
import android.content.Intent
import android.view.View
import android.widget.*
import com.sstechcanada.todo.activities.auth.LoginActivity
import com.google.android.gms.ads.AdView
import com.firebase.ui.firestore.FirestoreRecyclerOptions
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.gms.ads.AdRequest
import com.google.firebase.firestore.SetOptions
import es.dmoral.toasty.Toasty
import com.google.firebase.firestore.Query
import com.sstechcanada.todo.models.Category
import kotlinx.android.synthetic.main.act_bar.*
import kotlinx.android.synthetic.main.activity_category.*
import java.util.HashMap

class AddCategoryActivity : AppCompatActivity(),
    CategoryFirestoreAdapter.Callbacks {
    var userID: String? = null
    private var mAuth: FirebaseAuth? = null
    var db = FirebaseFirestore.getInstance()
    private var benefitCollectionRef: CollectionReference? = null
    private var userColRef: CollectionReference? = null
    private var categoryFirestoreAdapter: CategoryFirestoreAdapter? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_category)
        mAuth = FirebaseAuth.getInstance()
        userID = mAuth!!.currentUser!!.uid
        benefitCollectionRef = db.collection("Users").document(userID!!).collection("Benefits")
        userColRef = db.collection("Users").document(userID!!).collection("Lists")
        arrow_back.visibility = View.VISIBLE
        arrow_back.setOnClickListener { super.onBackPressed() }
        setUpRecyclerView()
        categoryFirestoreAdapter?.setCallbacks(this@AddCategoryActivity)
        toolbarTitle.text = getString(R.string.add_benefit)
        profile_toolbar.setOnClickListener {
            startActivity(
                Intent(this@AddCategoryActivity, LoginActivity::class.java)
            )
        }
        buttonAddCategory.setOnClickListener { addCategory() }
        if (MasterTodoListActivity.purchaseCode == "0") {
            val adView = findViewById<AdView>(R.id.adView)
            val adRequest = AdRequest.Builder().build()
            adView.loadAd(adRequest)
        }
    }

    private fun setUpRecyclerView() {
        val query: Query? = benefitCollectionRef
        val options = query?.let {
            FirestoreRecyclerOptions.Builder<Category>().setQuery(
                it, Category::class.java
            ).build()
        }
        categoryFirestoreAdapter = options?.let { CategoryFirestoreAdapter(it, this) }
        listViewCategory.setHasFixedSize(true)
        listViewCategory.layoutManager = LinearLayoutManager(this)
        listViewCategory.adapter = categoryFirestoreAdapter
    }

    override fun onStart() {
        super.onStart()
        categoryFirestoreAdapter!!.startListening()
    }

    override fun onStop() {
        super.onStop()
        categoryFirestoreAdapter!!.stopListening()
    }

    private fun addCategory() {
        //getting the values to save
        val name = editTextName!!.text.toString().trim { it <= ' ' }

        //checking if the value is provided
        if (name.isNotEmpty()) {
            val documentReferenceCurrentReference = benefitCollectionRef!!.document()
            val category: MutableMap<String, String> = HashMap()
            category["category_name"] = name
            documentReferenceCurrentReference.set(category, SetOptions.merge())
                .addOnSuccessListener {
                    Toasty.success(
                        applicationContext, "Benefit added", Toast.LENGTH_SHORT
                    ).show()
                }
                .addOnFailureListener {
                    Toasty.error(
                        applicationContext,
                        "Benefit addition Failed: ",
                        Toast.LENGTH_LONG
                    ).show()
                }

//            //Saving the Category
//            databaseCategories.child(id).setValue(category);

            //setting edittext to blank again
            editTextName!!.setText("")

            //displaying a success toast
        } else {
            //if the value is not given displaying a toast
            Toasty.warning(this, "Please enter a name", Toast.LENGTH_SHORT).show()
        }
    }

    override fun showProgressBar() {
        progressCat.visibility = View.VISIBLE
    }

    override fun hideProgressBar() {
        progressCat.visibility = View.INVISIBLE
    }

}